Single Choice and Multiple Choice Veto

JakeHartnell opened the pull request:
Closes #736

JakeHartnell committed:
Initial timelock / veto implementation

JakeHartnell committed:
Add refund logic for veto

JakeHartnell committed:
Refactor timelock

Should break less of the existing tests, doesn't lead to weird migration
state (passed at time zero), and if timelock configuration is updated it
won't affect currently open proposals.

JakeHartnell committed:
Start v2 migration

JakeHartnell committed:
Start writing tests

JakeHartnell committed:
Fix build and basic tests

JakeHartnell committed:
Fix veto checks, add more tests

JakeHartnell committed:
Cleanup unneeded changes

0xJayak commented on contracts/external/dao-migrator/Cargo.toml:
```diff
@@ -43,7 +44,13 @@ cw20-stake-v1                 = { workspace = true, features = ["library"] }
 cw4-voting-v1                 = { package = "cw4-voting", version = "0.1.0", git = "https://github.com/DA0-DA0/dao-contracts.git", tag = "v1.0.0" }
 cw20-v1                       = { version = "0.13", package = "cw20" }
-cw4-v1                       = { version = "0.13", package = "cw4" }
+cw4-v1                       =  { version = "0.13", package = "cw4" }
```
extra empty space

0xJayak commented on contracts/pre-propose/dao-pre-propose-approver/src/tests.rs:
```diff
@@ -14,7 +14,7 @@ use dao_pre_propose_approval_single::{
 };
 use dao_pre_propose_base::{error::PreProposeError, msg::DepositInfoResponse, state::Config};
-use dao_proposal_single as cps;
+use dao_proposal_single as dps;
```
I would put cps to dps change into a dedicated PR which will make this PR smaller and more readable. Happy to handle it if you don't mind.

JakeHartnell replied:
Please! A separate PR for this would be nice.

0xJayak replied:
Done, see #757

JakeHartnell replied:
Merged! Thank you!

0xJayak commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -255,6 +258,77 @@ pub fn execute_propose(
+    proposal_id: u64,
+) -> Result<Response, ContractError> {
+    let mut prop = PROPOSALS
+        .may_load(deps.storage, proposal_id)?
```
What's is the reason for `may_load` instead of `load` here? Should be better just end here if no proposal found no?

0xJayak replied:
Here is the a snippet which works for me. But I see the `may_load` pattern is the standard in the source code in the other places.

`let mut prop = PROPOSALS
        .load(deps.storage, proposal_id)
        .or(Err(ContractError::NoSuchProposal { id: proposal_id }))?;`

noahsaso replied:
seems like `may_load(..)?.ok_or(..)` does the same thing 🤷 just a style difference i guess

bekauz replied:
`may_load()` returns the data if it's there, none if it's not there, or an error if there are any parsing issues.

`load()` returns the data if it's there, or an error in case there is no data or there are any parsing issues.

in some cases it doesn't really matter, but in most cases it's nice to distinguish whether there is nothing stored from an error.

0xJayak commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -255,6 +258,77 @@ pub fn execute_propose(
+    match prop.status {
+        Status::Timelocked { expires } => {
+            match prop.timelock {
+                Some(ref timelock) => {
```
Alternatively, the nesting can be reduced with `if let`, e.g.:

`if let Status::Timelocked { expires } = prop.status {
    return Err(ContractError::NotPassed {});
}`

Which would also have a benefit of handling the error sooner.

0xJayak replied:
Sorry, got it wrong above, but could be refactored like this to reduce one level of nesting.

` let expires = match prop.status {
        Status::Timelocked { expires } => expires,
        _ => return Err(ContractError::NotPassed {})
    };`

JakeHartnell committed:
Veto before proposal passes feature (early veto)

Allows for the vetoer to veto a proposal before it passes is
veto_before_passed is set to true. Some DAOs may want this to be able to
speed up their process.

JakeHartnell committed:
Improve code reuse with proposal_completed_hooks

JakeHartnell committed:
Fix formatting

JakeHartnell committed:
Start fixing up migration tests

JakeHartnell committed:
dao-migrator test fixups

JakeHartnell committed:
Tests pass, made TODO notes on future work

JakeHartnell commented on packages/dao-voting/src/timelock.rs:
```diff
@@ -0,0 +1,102 @@
+#[cw_serde]
+pub struct Timelock {
+    /// The time duration to delay proposal execution for
+    pub delay: Timestamp,
```
Change to duration.

JakeHartnell commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -255,6 +256,111 @@ pub fn execute_propose(
+                    )?;
+
+                    // Add prepropose / deposit module hook which will handle deposit refunds.
+                    let proposal_creation_policy = CREATION_POLICY.load(deps.storage)?;
```
We should consider adding a new policy `NO_REFUND_ON_VETO`? Technically might be an extra case.

JakeHartnell replied:
Maybe rename. : )

bekauz replied:
I think it will be easier to add this after or with the prop-multi veto pr?

JakeHartnell replied:
Great.

JakeHartnell commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -255,6 +256,111 @@ pub fn execute_propose(
+                Some(ref timelock) => {
+                    // Check if the proposal timelock has expired, vetoer cannot veto
+                    // after expiration
+                    timelock.check_is_expired(env.block.time, expires)?;
```
Not sure if this warrants an option... open to ideas.

bekauz committed:
using Expiration & Duration for timelock; some cleanups

bekauz committed:
readme

bekauz committed:
unit tests wip

bekauz committed:
unit tests

bekauz committed:
fmt; test for vetoing non-open/timelocked props

bekauz committed:
adding semver dependency; v3 migration test assertions

bekauz commented:
@NoahSaso @JakeHartnell I think it's mostly ready. I'll give it another pass in a few days for the final cleanup.

One question though: do we want to accept votes when prop is `Timelocked` or `Vetoed`? atm voting is allowed until the prop `expiration` (not the timelock expiration but the prop itself) is active. Somewhat similar logic to how we allow voting when prop is passed early before expiration so that any members late for voting can still cast their votes to express their opinion.

JakeHartnell commented:
> One question though: do we want to accept votes when prop is Timelocked or Vetoed? atm voting is allowed until the prop expiration (not the timelock expiration but the prop itself) is active. Somewhat similar logic to how we allow voting when prop is passed early before expiration so that any members late for voting can still cast their votes to express their opinion.

Hmmmm, on the fence about this... guess we should make it consistent?

JakeHartnell commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -477,6 +481,7 @@ pub fn execute_vote(
     // cause a different one. This then serves to allow
     // for better tallies of opinions in the event that a
     // proposal passes or is rejected early.
+    // benskey TODO: check for potential timelock issues
```
TODO : )

JakeHartnell commented on contracts/proposal/dao-proposal-single/README.md:
```diff
@@ -59,3 +59,49 @@ handling a hook.
+Vetoer (`String`) is the address allowed to veto the proposals that are in
+`Timelocked` state.
+
+Vetoer address can be updated via a regular proposal config update.
```
Helpful note:
```
If you want the `vetoer` role to be shared between multiple organizations or individuals, a [cw1-whitelist](https://github.com/CosmWasm/cw-plus/tree/main/contracts/cw1-whitelist) contract address can be used to allow multiple accounts to veto the prop.
```

JakeHartnell commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -855,56 +968,53 @@ pub fn migrate(deps: DepsMut, _env: Env, msg: MigrateMsg) -> Result<Response, Co
             // blob we are migrating to. `version` is from storage. If
             // the version in storage matches the version in the blob
             // we are not upgrading.
+            // TODO fix this check!
```
Remove?

JakeHartnell commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -855,56 +968,53 @@ pub fn migrate(deps: DepsMut, _env: Env, msg: MigrateMsg) -> Result<Response, Co
+            let parsed_version = Version::from_str(&version)
+                .map_err(|_| ContractError::MigrationVersionError {})?;
+            // migration is only possible from 2.0.0
+            if parsed_version < Version::new(2, 0, 0) {
```
Also error if it's greater than the current version?

noahsaso commented:
> > One question though: do we want to accept votes when prop is Timelocked or Vetoed? atm voting is allowed until the prop expiration (not the timelock expiration but the prop itself) is active. Somewhat similar logic to how we allow voting when prop is passed early before expiration so that any members late for voting can still cast their votes to express their opinion.
> 
> Hmmmm, on the fence about this... guess we should make it consistent?

yeah i'm on the fence too.

on one hand, it feels weird that some may get to vote (and have their vote recorded/displayed) while others who did not vote before it was vetoed (pre-expiration) do not get to express an opinion.. but on the other hand, it feels meaningless to vote after a veto, so i foresee some people not voting while others vote anyway, leading to an outcome that is not representative of the members' opinions.

maybe it shouldn't be veto-able until after the prop is decided (i.e. it is early-passed because revoting is disabled, or it expires). or alternatively, votes could be deleted if it's vetoed early (that feels pretty weird too though). wdyt about that?

bekauz committed:
readme veto extension; validating migration from v2 versions

bekauz committed:
manual to_json_binary in pr scope

bekauz committed:
Merge branch 'development' into veto

bekauz committed:
lockfile

bekauz committed:
clippy fmt

bekauz committed:
happy clippy

bekauz committed:
schemas

bekauz committed:
test tube timelock fields

bekauz committed:
fmt

noahsaso commented on Cargo.toml:
```diff
@@ -116,13 +117,21 @@ dao-voting-cw721-staked = {path = "./contracts/voting/dao-voting-cw721-staked",
+stake-cw20-v03 = { package = "stake-cw20", version = "0.2.6" }
+
+# v2 dependencies. used for state migrations
+cw-utils-v2 = { package = "cw-utils", version = "0.16" }
+dao-dao-core-v2 = { package = "dao-dao-core", version = "2.2.0" }
+dao-interface-v2 = { package = "dao-interface", version = "2.2.0" }
+dao-proposal-single-v2 = { package = "dao-proposal-single", version = "2.2.0" }
+dao-pre-propose-single-v2 = { package = "dao-pre-propose-single", version = "2.2.0" }
+voting-v2 = { package = "dao-voting", version = "2.2.0" }
```
should we name these more explicitly (like `v220`)? surely we will have future migration dependencies that will require renaming these

JakeHartnell replied:
Migration needs serious thought... we need a good pattern if we're going to move fast.

noahsaso commented on contracts/external/dao-migrator/Cargo.toml:
```diff
@@ -27,13 +27,14 @@ cw20             = { workspace = true }
 dao-dao-core               = { workspace = true, features = ["library"] }
-dao-voting             = { workspace = true }
-dao-proposal-single    = { workspace = true, features = ["library"] }
+# dao-voting             = { workspace = true }
+# dao-proposal-single    = { workspace = true, features = ["library"] }
```
should these be removed?

JakeHartnell replied:
We need to think about a good migration pattern going forward... these are needed if we want the migrator to support `v1 -> v2`, `v2 -> v3`, etc... but in favor of removing them for now.

JakeHartnell replied:
Done.

noahsaso replied:
> We need to think about a good migration pattern going forward... these are needed if we want the migrator to support `v1 -> v2`, `v2 -> v3`, etc... but in favor of removing them for now.

well if they're for migrating then we should not remove them right?

noahsaso commented on contracts/external/dao-migrator/examples/schema.rs:
```diff
@@ -1,11 +1,10 @@
 use cosmwasm_schema::write_api;
-use dao_proposal_single::msg::{ExecuteMsg, InstantiateMsg, MigrateMsg, QueryMsg};
+use dao_migrator::msg::{ExecuteMsg, InstantiateMsg, QueryMsg};
```
lmao

noahsaso commented on contracts/proposal/dao-proposal-single/src/testing/contracts.rs:
```diff
@@ -39,16 +39,16 @@ pub(crate) fn cw20_stake_contract() -> Box<dyn Contract<Empty>> {
-    .with_migrate(cw_proposal_single_v1::contract::migrate);
-    Box::new(contract)
-}
+// pub(crate) fn v2_proposal_single_contract() -> Box<dyn Contract<Empty>> {
+//     let contract = ContractWrapper::new(
+//         dao_proposal_single_v2::contract::execute,
+//         dao_proposal_single_v2::contract::instantiate,
+//         dao_proposal_single_v2::contract::query,
+//     )
+//     .with_reply(dao_proposal_single_v2::contract::reply)
+//     .with_migrate(dao_proposal_single_v2::contract::migrate);
+//     Box::new(contract)
+// }
```
is this still needed?

JakeHartnell replied:
At some point it will be when we want to do contract migrations and tests, but let's kill it for now.

JakeHartnell replied:
Deleted.

noahsaso replied:
> At some point it will be when we want to do contract migrations and tests, but let's kill it for now.

wait we should definitely be doing contract migration and tests. i'm confused, we've been doing migrations and tests the whole time right? then we shouldn't delete this.

noahsaso commented on contracts/proposal/dao-proposal-single/src/state.rs:
```diff
@@ -55,6 +58,8 @@ pub struct Config {
     /// executed.
     pub close_proposal_on_execution_failure: bool,
+    /// Optional time delay on proposal execution
+    pub timelock: Option<Timelock>,
```
```suggestion
    /// Optional time delay on proposal execution during which the configured vetoer can veto the proposal.
    pub timelock: Option<Timelock>,
```

my initial thought was this comment could be improved, but i think `timelock` is just quite vague in general. what if we just call it `veto`? and the new status enum variant can be `VetoTimelock` instead of `Timelock`.

i think in general the `timelock` naming scheme could use a refactor to `veto` since that's really what it is

noahsaso commented on contracts/proposal/dao-proposal-single/src/proposal.rs:
```diff
@@ -61,14 +70,20 @@ impl SingleChoiceProposal {
+                },
+                // Otherwise the proposal is simply passed
+                None => Status::Passed,
+            },
```
this doesn't ever return `passed` if the configured timelock delay has expired. i feel like it would make sense that this switches to `passed` instead of returning `timelocked` with an `expiration` field in the past. currently this requires the expiration to be explicitly checked everywhere. on the other hand, leaving it as timelocked allows for more granular errors when attempting to execute or veto. wdyt?

noahsaso commented on contracts/proposal/dao-proposal-single/src/msg.rs:
```diff
@@ -199,26 +212,12 @@ pub enum QueryMsg {
+    FromV2 {
         /// This field was not present in DAO DAO v1. To migrate, a
         /// value must be specified.
         ///
```
comment no longer accurate for v2 timelock migration

noahsaso commented on contracts/proposal/dao-proposal-single/src/msg.rs:
```diff
@@ -199,26 +212,12 @@ pub enum QueryMsg {
 
 #[cw_serde]
 pub enum MigrateMsg {
-    FromV1 {
```
is there a reason to remove the v1 migration? is it difficult to support both?

noahsaso commented on contracts/proposal/dao-proposal-single/src/v2_state.rs:
```diff
@@ -0,0 +1,129 @@
+    match v2 {
+        voting_v2::status::Status::Open => Status::Open,
+        voting_v2::status::Status::Rejected => Status::Rejected,
+        voting_v2::status::Status::Passed => todo!(),
```
`todo!()` needs to be replaced

noahsaso commented on contracts/proposal/dao-proposal-single/src/v2_state.rs:
```diff
@@ -0,0 +1,129 @@
+        status_conversion!(Status::Rejected);
+        status_conversion!(Status::ExecutionFailed);
+        // TODO test passed status conversion
+        // status_conversion!(Status::Passed);
```
merp

noahsaso commented on packages/dao-voting/src/timelock.rs:
```diff
@@ -0,0 +1,73 @@
+    pub delay: Duration,
+    /// The address able to veto proposals.
+    pub vetoer: String,
+    /// Whether or not the vetoer can excute a proposal early before the
```
```suggestion
    /// Whether or not the vetoer can execute a proposal early before the
```

noahsaso commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -255,6 +258,114 @@ pub fn execute_propose(
+                        .add_submessages(proposal_completed_hooks))
+                }
+                // If timelock is not configured throw error
+                None => Err(ContractError::TimelockError(TimelockError::NoTimelock {})),
```
would it be better to check this one time at the beginning of the function instead of matching inside each status branch? if so, you could also move the `check_is_vetoer` calls to the top.

bekauz replied:
done! nice one

noahsaso commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -255,6 +258,114 @@ pub fn execute_propose(
+        .may_load(deps.storage, proposal_id)?
+        .ok_or(ContractError::NoSuchProposal { id: proposal_id })?;
+
+    match prop.status {
```
i believe this function is missing a `prop.update_status(&env.block);` call at the beginning before checking its status. the status stored in the proposal state may not be up to date based on the latest voting outcome if the voting period did not end early (like if revoting is allowed or quorum is not met). both the `execute` and `close` functions have this pattern where they update before reading the status, so i assume it is necessary here too.

for example: if the proposal is rejected due to lack of participation (quorum not met), this veto call will still see the proposal as open and try to veto it (if it has not yet been closed). calling `update_status` should ensure that it is up to date before it is checked.

bekauz replied:
good one, will add!

noahsaso commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -273,18 +384,43 @@ pub fn execute_execute(
             Some(prop.start_height),
         )?;
-        if power.is_zero() {
+        let vetoer_call = match config.timelock {
```
```suggestion
        let vetoer_call = match prop.timelock {
```

use `timelock` from proposal instead of config

JakeHartnell replied:
Very good catch.

JakeHartnell replied:
Very good catch.

noahsaso commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -273,18 +384,43 @@ pub fn execute_execute(
+    match prop.status {
+        Status::Passed => (),
+        Status::Timelocked { expiration } => {
+            // we can only end up in `Timelocked` state if `timelock`
+            // is not `None`
+            if let Some(ref timelock) = prop.timelock {
```
this comment is true, but everywhere else we check anyways and error if not, so might as well do it here to be insanely careful

JakeHartnell replied:
Agreed.

bekauz replied:
checking and returning an err now

noahsaso commented on packages/dao-hooks/src/all_hooks.rs:
```diff
@@ -14,6 +14,8 @@ pub enum DaoHooks {
     /// Called when NFTs are staked or unstaked.
     NftStakeChangeHook(NftStakeChangedHookMsg),
+    /// Pre-propose hooks
+    PreProposeHook(PreProposeHookMsg),
```
why is this being added here now? is it related to veto somehow?

JakeHartnell replied:
Consistency. "All hooks" should contain all hooks.

noahsaso commented on packages/dao-pre-propose-base/src/execute.rs:
```diff
@@ -286,19 +286,28 @@ where
+            && new_status != Status::Executed
+            && new_status != Status::Vetoed
+        {
             return Err(PreProposeError::NotClosedOrExecuted { status: new_status });
```
wdyt about updating this error to be `NotClosedExecutedOrVetoed`? or just `NotCompleted`? with a similar change in the error string of course

JakeHartnell replied:
I'm a fan of `NotCompleted`. You or @bekauz want to pick that up?

bekauz replied:
yep!

noahsaso reviewed (changes requested):
this is sooooo cool, super excited for this!!!!! and great job.

i have a handful of questions and a few bugs i believe need correcting. also, two discussion points (that are both covered by conversations i opened in this review, just surfacing them here for visibility):
1. i think we should rename timelock entirely to veto. timelock is ambiguous and somewhat distracting. i think it makes sense for the status to be `VetoTimelock` as that describes the state of a proposal, but the config option being called `veto` makes way more sense to me than `timelock`. thoughts?
2. should the status of proposals once the veto timelock expires revert back to passed? or should it stay timelocked with an expired expiration field? i'm not really sure. i mentioned some pros and cons in the conversation where i thought of it.

JakeHartnell commented:
> this is sooooo cool, super excited for this!!!!! and great job.
> 
> i have a handful of questions and a few bugs i believe need correcting. also, two discussion points (that are both covered by conversations i opened in this review, just surfacing them here for visibility):

Thanks for the awesome and helpful bug discovery and comments!

> 1. i think we should rename timelock entirely to veto. timelock is ambiguous and somewhat distracting. i think it makes sense for the status to be `VetoTimelock` as that describes the state of a proposal, but the config option being called `veto` makes way more sense to me than `timelock`. thoughts?

Yes. 100%.

> 2. should the status of proposals once the veto timelock expires revert back to passed? or should it stay timelocked with an expired expiration field? i'm not really sure. i mentioned some pros and cons in the conversation where i thought of it.

Hmmmmmm, this is a really good question... going to think about this one while addressing low hanging fruit.

JakeHartnell committed:
Bugfix: use timelock from proposal instead of config

Co-authored-by: noah <noahsaso@berkeley.edu>

JakeHartnell committed:
Better doc comment

Co-authored-by: noah <noahsaso@berkeley.edu>

Commit:
Cleanup unused code

bekauz committed:
checking validating vetoer early in veto; sanity check in execution from timelocked state

bekauz committed:
veto updating state before logic

bekauz committed:
pre propose error NotClosedOrExecuted renaming to NotCompleted

bekauz committed:
renaming Timelocked prop status to VetoTimelock; timelock rename to veto

bekauz committed:
replacing notion of timelock with veto config

bekauz committed:
just gen

JakeHartnell commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -855,56 +952,58 @@ pub fn migrate(deps: DepsMut, _env: Env, msg: MigrateMsg) -> Result<Response, Co
-        MigrateMsg::FromV1 {
-            close_proposal_on_execution_failure,
-            pre_propose_info,
-        } => {
```
Let's restore this.

JakeHartnell commented on contracts/proposal/dao-proposal-single/src/contract.rs:
```diff
@@ -855,56 +952,58 @@ pub fn migrate(deps: DepsMut, _env: Env, msg: MigrateMsg) -> Result<Response, Co
-        } => {
+        MigrateMsg::FromV2 { veto } => {
+            let version_2 = Version::new(2, 0, 0);
+
```
Let's kill this for now. Will support migrations at the next major version.

If a DAO really wants to use this, they can just swap out a new proposal module with veto from the previous one they were using.

bekauz committed:
revert v2 migration on prop-single

noahsaso commented on Cargo.toml:
```diff
@@ -116,13 +116,23 @@ dao-voting-cw721-staked = {path = "./contracts/voting/dao-voting-cw721-staked",
+stake-cw20-v03 = { package = "stake-cw20", version = "0.2.6" }
+
+
+# v2 dependencies. used for state migrations
+cw-utils-v2 = { package = "cw-utils", version = "0.16" }
+dao-dao-core-v2 = { package = "dao-dao-core", version = "2.2.0" }
+dao-interface-v2 = { package = "dao-interface", version = "2.2.0" }
+dao-proposal-single-v2 = { package = "dao-proposal-single", version = "2.2.0" }
+dao-pre-propose-single-v2 = { package = "dao-pre-propose-single", version = "2.2.0" }
+voting-v2 = { package = "dao-voting", version = "2.2.0"}
```
are these still needed now that we reverted the v2 migration?

noahsaso commented on contracts/external/dao-migrator/Cargo.toml:
```diff
@@ -43,7 +42,14 @@ cw20-stake-v1                 = { workspace = true, features = ["library"] }
-cw4-v1                       = { version = "0.13", package = "cw4" }
+cw4-v1                        = { version = "0.13", package = "cw4" }
+
+# v2 migration
+dao-dao-core-v2 = { workspace = true }
+dao-interface-v2 = { workspace = true }
+cw-utils-v2 = { workspace = true }
+dao-proposal-single-v2 = { workspace = true, features = ["library"] }
+voting-v2 = { workspace = true}
```
are these still needed now that we reverted the v2 migration?

noahsaso reviewed (changes requested):
sick so so close

noahsaso committed:
Reverted migrator changes and added veto to migration params.

noahsaso committed:
Updated dao-migrator schema.

noahsaso committed:
Ran clippy and formatter.

noahsaso committed:
Move proposal into passed state after the veto time lock expires.

noahsaso committed:
Updated schema.

noahsaso committed:
Fixed comments.

noahsaso committed:
Added veto to dao-proposal-multiple.

noahsaso committed:
Fixed schema again.

bekauz committed:
prop multi tests

bekauz committed:
validating veto config

bekauz committed:
lints

bekauz committed:
cleanup tests

noahsaso committed:
Fixed veto expiration calculation and removed timelock duration validation.

noahsaso committed:
Updated veto README and added to dao-proposal-multiple.

noahsaso committed:
Updated schema.

noahsaso committed:
Added validation that timelock_duration is the same units as the max_voting_period, and fixed tests. Also fixed missing timelock status handler.
